### DISCUSSION TITLE
Unimplemented pattern

### DISCUSSION THREAD

**Discussion by AlexisQapa** [pointfreeco/swift-composable-architecture#960] [Ideas]
Hi,

Playing around your latest attempts at updating the lib with concurrency, I remembered a nice tip from here : 
https://www.swiftbysundell.com/tips/using-function-for-userdefaults-key-consistency/

```swift
  static let failing = Self(
    searchLocation: { _ in throw Unimplemented(endpoint: #function) },
    weather: { _ in throw Unimplemented(endpoint: #function) }
  )
```

instead of

```swift
  static let failing = Self(
    searchLocation: { _ in throw Unimplemented(endpoint: "searchLocation") },
    weather: { _ in throw Unimplemented(endpoint: "weather") }
  )
```

This is nice but one need to properly name the struct fields. I guess a better solution could be made around Mirror or maybe simply by wrapping it in a free function with a shape like absurd.

edit: this surely is workable with a publisher too

**Comment by stephencelis** [comment]
Hey @AlexisQapa! Thanks for the suggestion, but we're not quite sure how this would work. Can you try sketching out a working implementation?

**Reply by AlexisQapa** [reply]
```swift
weather: { _ in throw Unimplemented(endpoint: #function) }
```
Is working as is, #function is replaced with "weather" at compile time. I'll give a try to wrap this into a function, my goal is something like

```swift
static let failing = Self(
    searchLocation: unimplemented(),
    weather: unimplemented(someOptionalReason)
  )
```

Thought I'm not sure I can make it work without multiple overloads (one per parameter count)

**Reply by iampatbrown** [reply]
@AlexisQapa do you mind sharing your full implementation?

For me using `#function` results in `Unimplemented(endpoint: "WeatherClient")`

**Reply by AlexisQapa** [reply]
Indeed, I could swear I got `Unimplemented(endpoint: "searchLocation")` yesterday. Well this all falls apart now :/

**Reply by iampatbrown** [reply]
It's an interesting idea though :) I think the main difference with this and the user defaults example you provided was that `#function` is being called from a computed property.

**Reply by AlexisQapa** [reply]
Yes, #function match the struct initialiser.

For those who use protocols instead of structs for the env : 
```swift
protocol WeatherClientProtocol {

	func searchLocation(_ query: String) async throws -> [Location]
	func weather(_ id: Int) async throws -> LocationWeather
}

extension WeatherClientProtocol {
	func searchLocation(_ query: String) async throws -> [Location] {
		throw Unimplemented(endpoint: #function)
	}

	func weather(_ id: Int) async throws -> LocationWeather {
		throw Unimplemented(endpoint: #function)
	}
}
```
Using protocols I get: 
```
TaskResult.failure(Unimplemented(endpoint: "searchLocation(_:)"))
```
